sdks/java/harness/src/test/java/org/apache/beam/fn/harness/state/StateBackedIterableTest.java

```diff
@@ -171,6 +173,38 @@ public void testEncodeDecodeStateBackedIterable() throws Exception {
       assertEquals(iterable.prefix, result.prefix);
       assertEquals(iterable.request, result.request);
     }
+
+    @Test
+    public void testSerializability() throws Exception {
+      FakeBeamFnStateClient fakeBeamFnStateClient =
+          new FakeBeamFnStateClient(
+              ImmutableMap.of(
+                  key("suffix"), encode("C", "D", "E"),
+                  key("emptySuffix"), encode()));
+
+      StateBackedIterable<String> iterable =
+          new StateBackedIterable<>(
+              fakeBeamFnStateClient,
+              "instruction",
+              encode("suffix"),
+              StringUtf8Coder.of(),
+              ImmutableList.of("A", "B"));
+
+      List<String> expected = ImmutableList.of("A", "B", "C", "D", "E");
+
+      ByteArrayOutputStream bout = new ByteArrayOutputStream();
+      ObjectOutputStream out = new ObjectOutputStream(bout);
+      out.writeObject(iterable);
+      out.flush();
+      ByteArrayInputStream bin = new ByteArrayInputStream(bout.toByteArray());
+      ObjectInputStream in = new ObjectInputStream(bin);
+      Iterable<String> deserialized = (Iterable<String>) in.readObject();
```

Comment:
Probably a dumb question. The `beamFnStateClient` field is set to transient. How could the deserialized iterable keep the information from `fakeBeamFnStateClient`?

Reply:
We never actually serialize a `StateBackedIterable` due to the `writeReplace` method. Instead, whenever we are asked to serialize a `StateBackedIterable` we tell Java to serialize a materialized copy of ourselves instead. 

`beamFnStateClient` is marked transient just to make spotBugs happy.

Reply:
Oh, I see. `writeReplace` is a special method for serializable objects. Thanks for the clarification.